crates/revm/src/evm_impl.rs

```diff
@@ -190,18 +190,25 @@ impl<'a, GSPEC: Spec, DB: Database, const INSPECT: bool> Transact<DB::Error>
         let effective_gas_price = env.effective_gas_price();
 
         #[cfg(feature = "optimism")]
-        let (tx_mint, tx_system, is_deposit) = (
-            env.tx.optimism.mint,
-            env.tx.optimism.is_system_transaction,
-            env.tx.optimism.source_hash.is_some(),
-        );
+        let (tx_mint, tx_system, tx_l1_cost, is_deposit, l1_block_info) = {
+            let is_deposit = env.tx.optimism.source_hash.is_some();
 
-        // Perform this calculation optimistically to avoid cloning the enveloped tx.
-        #[cfg(feature = "optimism")]
-        let tx_l1_cost = {
             let l1_block_info =
-                optimism::L1BlockInfo::try_fetch(&mut self.data.db).map_err(EVMError::Database)?;
-            l1_block_info.calculate_tx_l1_cost::<GSPEC>(&env.tx.optimism.enveloped_tx, is_deposit)
+                optimism::L1BlockInfo::try_fetch(self.data.db, self.data.env.cfg.optimism)?;
+
+            // Perform this calculation optimistically to avoid cloning the enveloped tx.
+            let tx_l1_cost = l1_block_info.clone().map(|l1_block_info| {
```

Comment:
Why do we need to clone?

Reply:
Shouldn't we be able to do
```rust 
let tx_l1_cost = l1_block_info.map(|block_info|
    block_info.calculate_tx_l1_cost::<GSPEC>(&env.tx.optimism.enveloped_tx, is_deposit)
);
```

Reply:
`.map` accepts an `FnOnce`, which consumes `T`

Reply:
Not super desirable, but given that the l1 block info is very small, clone should be low-cost

Reply:
You can just take the option as a reference like so:
```suggestion
            let tx_l1_cost = l1_block_info.as_ref().map(|l1_block_info| {
```